perception/lidar_transfusion/lib/transfusion_trt.cpp

```diff
@@ -0,0 +1,262 @@
+
+  params_input_d_ = cuda::make_unique<unsigned int>();
+  voxel_features_d_ = cuda::make_unique<float[]>(voxel_features_size_);
+  voxel_num_d_ = cuda::make_unique<unsigned int[]>(voxel_num_size_);
+  voxel_idxs_d_ = cuda::make_unique<unsigned int[]>(voxel_idxs_size_);
+  points_d_ = cuda::make_unique<float[]>(config_.cloud_capacity_ * config_.num_point_feature_size_);
+  pre_ptr_.reset(new PreprocessCuda(config_, stream_));
+  post_ptr_.reset(new PostprocessCuda(config_, stream_));
+}
+
+bool TransfusionTRT::detect(
+  const sensor_msgs::msg::PointCloud2 & msg, const tf2_ros::Buffer & tf_buffer,
+  std::vector<Box3D> & det_boxes3d, std::unordered_map<std::string, double> & proc_timing)
+{
+  stop_watch_ptr_->toc("processing/inner", true);
+  if (!preprocess(msg, tf_buffer)) {
+    RCLCPP_WARN_STREAM(
+      rclcpp::get_logger("lidar_transfusion"), "Fail to preprocess and skip to detect.");
+    return false;
+  }
+  proc_timing.emplace(
+    "debug/processing_time/preprocess_ms", stop_watch_ptr_->toc("processing/inner", true));
+
+  if (!inference()) {
+    RCLCPP_WARN_STREAM(
+      rclcpp::get_logger("lidar_transfusion"), "Fail to inference and skip to detect.");
+    return false;
+  }
+  proc_timing.emplace(
+    "debug/processing_time/inference_ms", stop_watch_ptr_->toc("processing/inner", true));
+
+  if (false) {
+#ifdef CPU_PROCESSING
+  } else if (!postprocessCPU(det_boxes3d)) {
+#else
+  } else if (!postprocess(det_boxes3d)) {
+#endif
+    RCLCPP_WARN_STREAM(
+      rclcpp::get_logger("lidar_transfusion"), "Fail to postprocess and skip to detect.");
+    return false;
+  }
+  proc_timing.emplace(
+    "debug/processing_time/postprocess_ms", stop_watch_ptr_->toc("processing/inner", true));
+
+  return true;
+}
+
+bool TransfusionTRT::preprocess(
+  const sensor_msgs::msg::PointCloud2 & msg, const tf2_ros::Buffer & tf_buffer)
+{
+  if (!vg_ptr_->enqueuePointCloud(msg, tf_buffer)) {
+    return false;
+  }
+
+  cuda::clear_async(cls_output_d_.get(), cls_size_, stream_);
+  cuda::clear_async(box_output_d_.get(), box_size_, stream_);
+  cuda::clear_async(dir_cls_output_d_.get(), dir_cls_size_, stream_);
+  cuda::clear_async(voxel_features_d_.get(), voxel_features_size_, stream_);
+  cuda::clear_async(voxel_num_d_.get(), voxel_num_size_, stream_);
+  cuda::clear_async(voxel_idxs_d_.get(), voxel_idxs_size_, stream_);
+  cuda::clear_async(params_input_d_.get(), 1, stream_);
+  cuda::clear_async(
+    points_d_.get(), config_.cloud_capacity_ * config_.num_point_feature_size_, stream_);
+
+  const auto count = vg_ptr_->generateSweepPoints(msg, points_d_);
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("lidar_transfusion"), "Generated sweep points: " << count);
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+
+  pre_ptr_->generateVoxels(
+    points_d_.get(), count, params_input_d_.get(), voxel_features_d_.get(), voxel_num_d_.get(),
+    voxel_idxs_d_.get());
+  unsigned int params_input_cpu;
```

Comment:
The nomenclature is usually host/device, rather than cpu/gpu

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/389c6cbfc28ff9abf4be285029f6a19a9101ed1b